app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt

```diff
@@ -419,6 +449,480 @@ class Downloader(
+            try {
+                return executeChapterDownload(download, mangaDir, chapterDirname)
+            } catch (e: Exception) {
+                lastException = e
+
+                // Check if this exception should be retried
+                if (!shouldRetryException(e)) {
+                    logcat(LogPriority.ERROR) {
+                        "Full chapter download failed with non-retryable error: ${e.message}"
+                    }
+                    throw e
+                }
+
+                retryCount++
+
+                if (retryCount <= MAX_DOWNLOAD_RETRIES) {
+                    logcat(LogPriority.WARN) {
+                        "Full chapter download failed (attempt $retryCount/$MAX_DOWNLOAD_RETRIES): ${e.message}"
+                    }
+                    delay(RETRY_DELAY_MS * retryCount)
+                }
+            }
+        }
+
+        throw lastException ?: IOException("Download failed after $MAX_DOWNLOAD_RETRIES retries")
+    }
+
+    /**
+     * Executes a single chapter download attempt.
+     *
+     * @param download the chapter download
+     * @param mangaDir the manga directory
+     * @param chapterDirname the chapter directory name
+     * @return the downloaded chapter file
+     */
+    private suspend fun executeChapterDownload(
+        download: Download,
+        mangaDir: UniFile,
+        chapterDirname: String,
+    ): UniFile {
+        val chapterFile = createChapterFile(mangaDir, chapterDirname)
+
+        return download.source.getFullChapterResponse(download.chapter.toSChapter()).use { response ->
+            streamResponseToFile(response, chapterFile, download)
+
+            // Validate the downloaded file
+            validateDownloadedFile(chapterFile)
+
+            chapterFile
+        }
+    }
+
+    /**
+     * Creates and validates the chapter file.
+     *
+     * @param mangaDir the manga directory
+     * @param chapterDirname the chapter directory name
+     * @return the created chapter file
+     */
+    private fun createChapterFile(mangaDir: UniFile, chapterDirname: String): UniFile {
+        if (!mangaDir.exists()) {
+            throw IOException("Manga directory does not exist: ${mangaDir.uri}")
+        }
+
+        val chapterFile = mangaDir.createFile("$chapterDirname.cbz")
+            ?: throw IOException("Failed to create chapter file: $chapterDirname.cbz in directory: ${mangaDir.uri}")
+
+        if (!chapterFile.exists()) {
+            throw IOException("Chapter file was not created successfully: ${chapterFile.uri}")
+        }
+
+        return chapterFile
+    }
+
+    /**
+     * Streams the response content to the chapter file with progress tracking.
+     *
+     * @param response the HTTP response
+     * @param chapterFile the target file
+     * @param download the download for progress tracking
+     */
+    private suspend fun streamResponseToFile(response: Response, chapterFile: UniFile, download: Download) {
```

Comment:
This seems to provide another progress tracking approach different from what's used in `getImage()`.

https://github.com/mihonapp/mihon/blob/83f4b486296f6998efc8ef95fa9d58b75e5de130/source-api/src/commonMain/kotlin/eu/kanade/tachiyomi/source/online/HttpSource.kt#L374-L377

We're hesitant about making this `open` because the progress tracking will be broken if this is overridden. Maybe this method can be extracted to apply to that too?